src/librbd/ManagedLock.cc

```diff
@@ -0,0 +1,501 @@
+}
+
+template <typename W>
+void ManagedLock<W>::execute_action(Action action, Context *ctx) {
+  assert(m_lock.is_locked());
+
+  append_context(action, ctx);
+  if (!is_transition_state()) {
+    execute_next_action();
+  }
+}
+
+template <typename W>
+void ManagedLock<W>::execute_next_action() {
+  assert(m_lock.is_locked());
+  assert(!m_actions_contexts.empty());
+  switch (get_active_action()) {
+  case ACTION_ACQUIRE_LOCK:
+    send_acquire_lock();
+    break;
+  case ACTION_REACQUIRE_LOCK:
+    send_reacquire_lock();
+    break;
+  case ACTION_RELEASE_LOCK:
+    send_release_lock();
+    break;
+  case ACTION_SHUT_DOWN:
+    send_shutdown();
+    break;
+  default:
+    assert(false);
+    break;
+  }
+}
+
+template <typename W>
+typename ManagedLock<W>::Action ManagedLock<W>::get_active_action() const {
+  assert(m_lock.is_locked());
+  assert(!m_actions_contexts.empty());
+  return m_actions_contexts.front().first;
+}
+
+template <typename W>
+void ManagedLock<W>::complete_active_action(State next_state, int r) {
+  assert(m_lock.is_locked());
+  assert(!m_actions_contexts.empty());
+
+  ActionContexts action_contexts(std::move(m_actions_contexts.front()));
+  m_actions_contexts.pop_front();
+  m_state = next_state;
+
+  m_lock.Unlock();
+  for (auto ctx : action_contexts.second) {
+    ctx->complete(r);
+  }
+  m_lock.Lock();
+
+  if (!is_transition_state() && !m_actions_contexts.empty()) {
+    execute_next_action();
+  }
+}
+
+template <typename W>
+bool ManagedLock<W>::is_shutdown_locked() const {
+  assert(m_lock.is_locked());
+
+  return ((m_state == STATE_SHUTDOWN) ||
+          (!m_actions_contexts.empty() &&
+           m_actions_contexts.back().first == ACTION_SHUT_DOWN));
+}
+
+template <typename W>
+void ManagedLock<W>::send_acquire_lock() {
+  assert(m_lock.is_locked());
+  if (m_state == STATE_LOCKED) {
+    complete_active_action(STATE_LOCKED, 0);
+    return;
+  }
+
+  ldout(m_cct, 10) << this << " " << __func__ << dendl;
+  m_state = STATE_ACQUIRING;
+
+  assert(m_watcher->is_registered());
+  uint64_t watch_handle = m_watcher->get_watch_handle();
+
+  m_cookie = ManagedLock<W>::encode_lock_cookie(watch_handle);
+  AcquireRequest<W>* req = AcquireRequest<W>::create(this,
+    util::create_context_callback<
+      ManagedLock<W>, &ManagedLock<W>::handle_acquire_lock>(this));
+  m_work_queue->queue(new C_SendRequest<AcquireRequest<W>>(req), 0);
+}
+
+template <typename W>
+void ManagedLock<W>::handle_acquire_lock(int r) {
+  ldout(m_cct, 10) << this << " " << __func__ << ": r=" << r << dendl;
+
+  if (r == -EBUSY || r == -EAGAIN) {
+    ldout(m_cct, 5) << "unable to acquire exclusive lock" << dendl;
+  } else if (r < 0) {
+    lderr(m_cct) << "failed to acquire exclusive lock:" << cpp_strerror(r)
+               << dendl;
+  } else {
+    ldout(m_cct, 5) << "successfully acquired exclusive lock" << dendl;
+  }
+
+  State next_state = (r < 0 ? STATE_UNLOCKED : STATE_LOCKED);
+  if (r == -EAGAIN) {
```

Comment:
I would pass this error down to ```ExclusiveLock``` since it has the logic to (potentially) request the lock from a peer if it's in-use by an alive peer. This state machine should be interlocked w/ ```ExclusiveLock``` anyway so ```ExclusiveLock::handle_acquire_lock``` shouldn't need to check whether or not the lock was acquired since the error code will be the status.